3111 remove loop var from extraction

hiker opened the pull request:
This needs #2049 / #3131 to be merged in. It has a lot of clashes with 2049, so I've merged in 2049 here and resolved them.

arporter reviewed (changes requested):
Thanks @hiker, I think it's looking good.  Unfortunately, this is going to clash with #3164 which is removing the ComponentIndices class. I've therefore paused my review. The only real concern I have so far is with naming - I think some changes would help with readability.

hiker commented on src/psyclone/core/variables_access_map.py:
```diff
@@ -63,7 +66,12 @@ def __str__(self):
+            signature: Signature,
+            access_type: AccessType,
+            node: "Node",
+            component_indices: Optional[ComponentIndices] = None) -> None:
```
I temporarily merged this with #3164, and besides a few conflicts, and besides missing coverage for a line or two (as in: 100% coverage by the associated single test file) it's all good. I'll start working on your comments next.

hiker commented on src/psyclone/core/access_sequence.py:
```diff
@@ -61,21 +63,21 @@ class AccessInfo():
     '''
     def __init__(
-        self, access_type: AccessType, node: 'Node',
+        self, access_type: AccessType, node: Union['Node', Symbol],
```
Yes.

hiker commented on src/psyclone/psyir/nodes/extract_node.py:
```diff
@@ -49,21 +49,27 @@
+    ArrayType, ContainerSymbol, DataSymbol, DataType, ImportInterface,
+    INTEGER_TYPE, REAL8_TYPE, Symbol, SymbolTable)
+
+if TYPE_CHECKING:
```
That's not specific to `MyPy`, `TYPE_CHECKING` is standard python and supported by many (all?) type checking tools

arporter replied:
Oh! Every day is a school day...

hiker commented on src/psyclone/psyir/nodes/extract_node.py:
```diff
@@ -158,21 +164,79 @@ def extract_body(self):
         '''
         return self._post_name
 
+    def get_removable_variables(self) -> list[tuple[str, Signature]]:
```
Done.

hiker commented on src/psyclone/psyir/nodes/extract_node.py:
```diff
@@ -158,21 +164,79 @@ def extract_body(self):
+        # discarded (e.g. a loop variable might be declared as omp private).
+        # We do this by counting how many directive nodes are in the list, and
+        # then subtracting this number from all accesses:
+        for var_sig, accesses in all_loops.items():
```
OK, I've left it all the way it is, easy :)

hiker commented on src/psyclone/psyir/nodes/extract_node.py:
```diff
@@ -193,16 +257,15 @@ def lower_to_language_level(self):
         ctu = CallTreeUtils()
         read_write_info = ctu.get_in_out_parameters(
             self, include_non_data_accesses=False)
+        removable_vars = self.get_removable_variables()
```
Done.

hiker commented on src/psyclone/psyir/nodes/extract_node.py:
```diff
@@ -298,12 +371,12 @@ def _flatten_signature(signature: Signature) -> str:
 
         '''
-        already_flattened = {}  # dict of name: symbol
+        already_flattened: dict[str, Symbol] = {}  # dict of name: symbol
```
Not too bad ;) But see my final comment later.

arporter replied:
Phew :-)

hiker commented:
I have addressed all issues, will trigger CI and then mark this as ready for review, but I am happy to with for #3164 to be merged in first (and I then sort out the few conflicts, which are not too bad), given that #3164 also blocks my training PR to go in.

arporter commented on src/psyclone/psyir/tools/read_write_info.py:
```diff
@@ -120,7 +123,9 @@ def all_used_vars_list(self):
+                 container_name: Optional[str] = None) -> None:
         '''This function adds a read access to the specified signature and
         container name. The container_name is optional and defaults to "",
         indicating that this signature is not based on importing a symbol
```
Please rm the `:type ...` parts of this docstring.

hiker replied:
Done.

arporter commented on src/psyclone/psyir/tools/read_write_info.py:
```diff
@@ -139,7 +144,9 @@ def add_read(self, signature, container_name=None):
+                  container_name: Optional[str] = None) -> None:
         '''This function adds a write access to the specified signature and
         container name. The container_name is optional and defaults to "",
         indicating that this signature is not based on importing a symbol
```
Please rm the `:type ..` parts below.

arporter commented on src/psyclone/psyir/tools/read_write_info.py:
```diff
@@ -158,10 +165,47 @@ def add_write(self, signature, container_name=None):
+        '''
+        Checks if the given signature is in the read list.
+
+        :param signature: the signature to remove
```
s/remove/check/

arporter commented on src/psyclone/psyir/tools/read_write_info.py:
```diff
@@ -158,10 +165,47 @@ def add_write(self, signature, container_name=None):
+    # -------------------------------------------------------------------------
+    def remove(self, signature: Signature, container_name: str = "") -> None:
+        '''
+        This function removes a signature (if required with the corresponding
```
Is `container_name` Optional? If so pls update the typehint and explain when container_name should be provided and what happens if it isn't.

hiker replied:
Indeed. Updated the typing and comments.

arporter commented on src/psyclone/tests/domain/gocean/transformations/gocean_extract_test.py:
```diff
@@ -504,9 +483,8 @@ def test_driver_generation_flag(create_driver):
-    fails because of #641.
-
+    '''Test that loop variables are not stored, and also not
+    read in the driver.
```
Whoop :rocket:

arporter commented on src/psyclone/tests/domain/lfric/transformations/lfric_extract_test.py:
```diff
@@ -908,16 +877,18 @@ def test_extract_colouring_omp_lfric():
+            in code)
+
+    # Verify that loop variables are not extracted:
+    assert ('CALL extract_psy_data % ProvideVariable("cell_post", cell)'
```
Please make more robust if poss.

arporter commented on src/psyclone/tests/domain/lfric/transformations/lfric_extract_test.py:
```diff
@@ -748,7 +723,7 @@ def test_extract_kernel_and_builtin_lfric():
 
 
-def test_extract_colouring_omp_lfric():
+def test_extract_colouring_omp_lfric(tmpdir):
```
I can't see how the `tmpdir` fixture is being used?

hiker replied:
Oh yes, I tried to enable compilation, but that failed (because the extract library is not compiled, otherwise it all works). Off topic for this PR, so I've removed tmdir again.

hiker commented on src/psyclone/tests/core/access_sequence_test.py:
```diff
@@ -43,12 +43,13 @@
 
 
-def test_access_info():
+def test_access_info() -> None:
```
I thought that (at least) mypy does not check any functions that have no typing in their declaration. So, that was why I started to add it. But when I tried again, I couldn't reproduce this behaviour anymore, no idea if I made a mistake in the past, or if mypy has changed.
Technically, adding '->None' is considered to be best practice (to ensure that it's obvious that the function is typed, and not that it was forgotten), but as long as our (future) tool checks this, I actually don't care. I've left them in place, but did not add more.

hiker commented on src/psyclone/tests/psyir/nodes/extract_node_test.py:
```diff
@@ -495,3 +500,123 @@ def test_extraction_flatten_datatype(monkeypatch):
+
+    etrans = GOceanExtractTrans()
+
+    # Test a Loop nested within the OMP Parallel DO Directive
```
Indeed ... nor is there one in the other two tests that have the same comment :) I've removed the comment, I believe the tests are fine. Likely a very old copy&paste left-over.

arporter commented on src/psyclone/tests/domain/common/driver_creator_test.py:
```diff
@@ -138,11 +141,14 @@ def test_lfric_driver_simple_test():
-            not in driver.lower())
+        assert line.lower() in driver, line
+
+    print("XX", driver)
```
Done.
